### PR TITLE
[#25] Add parse!/2 and parse_params!/2

### LINKED ISSUE #25: Streamline Usage with `parse_params!` and `parse!`
If an error happens with either of these calls, they will indicate in the `Filtrex` struct that the filter is empty:

``` elixir
%Filtrex{empty: true}
```

When this value is piped into `Filtrex.query/2`, it will by default not narrow the query but could be configurable to always return an empty set of values:

``` elixir
# default behavior:
Filters.query(YourApp.YourModel, %Filtrex{empty: true})
# => does not narrow query

# optional behavior
Filters.query(YourApp.YourModel, %Filtrex{empty: true}, allow_empty: false)
# => returns query that never matches (e.g. true == false or something similar)
```

### PR DESCRIPTION
**Pull request by danielberkompas**
Closes #25, built according to the spec described there.

### CODE REVIEW

File: lib/filtrex.ex [hunk 1]
```diff
@@ -43,29 +43,75 @@ defmodule Filtrex do
...
+  set the `allow_empty` option to `true`:
+
+  ```
+  Filtrex.query(query, filter, allow_empty: true)
+  ```
   """
-  @spec query(Ecto.Query.t, Filtrex.t) :: Ecto.Query.t
-  def query(query, %Filtrex{empty: true}), do: query
+  @spec query(Ecto.Query.t, Filtrex.t, Keyword.t) :: Ecto.Query.t
+  def query(query, filter, opts \\ [allow_empty: true])
```
Comment (credoci) [inline lib/filtrex.ex]: Parameter `query` has same name as a function in the same module.

File: lib/filtrex.ex [hunk 2]
```diff
@@ -43,29 +43,75 @@ defmodule Filtrex do
...
+  def query(query, %Filtrex{empty: true}, opts) do
```
Comment (credoci) [inline lib/filtrex.ex]: Parameter `query` has same name as a function in the same module.

File: lib/filtrex.ex [hunk 3]
```diff
@@ -43,29 +43,75 @@ defmodule Filtrex do
...
+    if opts[:allow_empty] do
+      query
+    else
+      Ecto.Query.where(query, false)
+    end
+  end
 
-  def query(query, filter) do
-    {result, _} = Filtrex.AST.build_query(query, filter)
+  def query(query, filter, _opts) do
```
Comment (credoci) [inline lib/filtrex.ex]: Parameter `query` has same name as a function in the same module.

**Comment by rcdilorenzo** [comment]
@danielberkompas, thanks so much! I'll try to take a look at tonight or tomorrow. I really appreciate it. BTW, what library are you building (if you don't mind me asking)?

**Comment by danielberkompas** [comment]
@rcdilorenzo https://hexdocs.pm/torch/1.0.0-rc.1/readme.html

It's not quite out yet, but the docs are up and you can try it.

File: lib/filtrex.ex [hunk 4]
```diff
@@ -43,29 +43,75 @@ defmodule Filtrex do
...
+  set the `allow_empty` option to `true`:
+
+  ```
+  Filtrex.query(query, filter, allow_empty: true)
+  ```
   """
-  @spec query(Ecto.Query.t, Filtrex.t) :: Ecto.Query.t
-  def query(query, %Filtrex{empty: true}), do: query
+  @spec query(Ecto.Query.t, Filtrex.t, Keyword.t) :: Ecto.Query.t
+  def query(queryable, filter, opts \\ [allow_empty: true])
```
Comment (rcdilorenzo) [inline lib/filtrex.ex:102]: I didn't know that you can do this to define default options. Learn something new every day!

**Pull request merged**